Sources/_StringProcessing/RegexDSL/Core.swift

```diff
@@ -13,39 +11,47 @@ public protocol RegexProtocol {
   var regex: Regex<Capture> { get }
 }
 
-public protocol EmptyProtocol {}
-public struct Empty: EmptyProtocol {}
-extension Array: EmptyProtocol where Element: EmptyProtocol {}
-extension Optional: EmptyProtocol where Wrapped: EmptyProtocol {}
-
 /// A regular expression.
 public struct Regex<Capture>: RegexProtocol {
   /// A program representation that caches any lowered representation for
   /// execution.
   internal class Program {
-    /// The underlying AST.
-    let ast: AST
+    /// The underlying pattern tree
+    let tree: PatternTree
+
+    var regexAST: AST? { tree.regexAST }
+
     /// The legacy `RECode` for execution with a legacy VM.
     lazy private(set) var legacyLoweredProgram: RECode = {
       do {
+        guard let ast = regexAST else {
+          fatalError("FIXME")
+        }
         return try compile(ast)
       } catch {
         fatalError("Regex engine internal error: \(String(describing: error))")
       }
     }()
     /// The program for execution with the matching engine.
-    lazy private(set) var loweredProgram = Compiler(ast: ast).emit()
+    lazy private(set) var loweredProgram = Compiler(
+      ast: regexAST! // FIXME
+    ).emit()
 
-    init(ast: AST) {
-      self.ast = ast
+    init(_ tree: PatternTree) {
```

Comment:
Nit: I’d prefer keeping an `ast:` or `tree:` argument label since it’s not exactly a value-preserving conversion and it’s better to be able to tell this is calling an internal API at call sites.

Reply:
When might it not be value-preserving? It's weird to have the type name as the label, if there's a value loss, is there a way to describe what could be lost?

Reply:
I imagined there would eventually be invalid trees? If so it would not be a monomorphism, otherwise never mind the value-preserving conversion argument.

That said, IMO the main value of having an argument label is to distinguish between calls to internal API and calls to public API more easily in our tests. I'm imagining there will be call sites like `Pattern(.hexDigit)`, where which API being called would be less distinguishable between that and `Pattern(.regex(...))`. An argument label will describe the argument's role and surely doesn't have to repeat the type information. Something like `init(representing:)` would work, too.